src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs

```diff
@@ -0,0 +1,89 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Features;
+    using Logging;
+    using Settings;
+    using SimpleJson;
+
+    class HostStartupDiagnostics : Feature
+    {
+        public HostStartupDiagnostics()
+        {
+            EnableByDefault();
+        }
+
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            var settings = context.Settings;
+
+            var diagnosticsWriter = GetDiagnosticsWriter(settings);
+
+            context.RegisterStartupTask(new WriteStartupDiagnostics(diagnosticsWriter, settings));
+        }
+
+
+        static HostDiagnosticsWriter GetDiagnosticsWriter(ReadOnlySettings settings)
+        {
+            if (settings.TryGet<HostDiagnosticsWriter>(out var diagnosticsWriter))
+            {
+                return diagnosticsWriter;
+            }
+
+            if (!settings.TryGet<string>(HostDiagnosticsConfigurationExtensions.DiagnosticsRootPathKey, out var diagnosticsRootPath))
+            {
+                diagnosticsRootPath = Path.Combine(Host.GetOutputDirectory(), ".diagnostics");
+            }
+
+            if (!Directory.Exists(diagnosticsRootPath))
+            {
+                Directory.CreateDirectory(diagnosticsRootPath);
+            }
+
+            var endpointName = settings.EndpointName();
+
+            // Once we have the proper hosting model in place we can skip the endpoint name since the host would
+            // know how to handle multi hosting but for now we do this so that multi-hosting users will get a file per endpoint
+            var startupDiagnosticsFileName = $"{endpointName}-configuration.json";
```

Comment:
This naming scheme means that we'll only ever have a single file. Should we consider some sort of multiple file/rolling scheme so that we can collect data for the last N endpoint starts?

Reply:
That was considered but scoped out for now since its really only the latest file we're interested in. Writing to disk is the breaking change, we can always improve on this by doing rolling files, hashes etc in a minor